server/api/controllers/device.controller.js

```diff
@@ -91,6 +92,129 @@ module.exports = function DeviceController(gladys) {
+   * @apiGroup Device
+   */
+  async function getDeviceFeature(req, res) {
+    const params = Object.assign({}, req.query, {
+      device_feature_selector: req.params.device_feature_selector.split(','),
+    });
+
+    // Choose attributes
+    params.attributes_device = [];
+    params.attributes_device.push('name');
+    params.attributes_device.push('selector');
+    params.attributes_device_feature = [];
+    params.attributes_device_feature.push('name');
+    params.attributes_device_feature.push('selector');
+    params.attributes_device_feature.push('unit');
+    params.attributes_device_feature.push('last_value');
+    params.attributes_device_feature.push('last_value_changed');
+    params.attributes_device_feature.push('last_value_changed');
+    params.attributes_device_room = [];
+    params.attributes_device_room.push('name');
+    params.attributes_device_room.push('selector');
+    params.attributes_device_service = [];
+    params.attributes_device_service.push('name');
+    params.attributes_device_param = [];
+    params.attributes_device_param.push('name');
+
+    const devices = await gladys.device.get(params);
+
+    res.json(devices);
+  }
+  /**
+   * @api {get} /api/v1/device_feature_sate/:device_feature_selector getDeviceFeatureStates
+   * @apiName getDeviceFeatureStates
+   * @apiGroup Device
+   */
+  async function getDeviceFeatureStates(req, res) {
+    const params = Object.assign({}, req.query, {
+      device_feature_selector: req.params.device_feature_selector.split(','),
+    });
+
+    // fix period
+    const beginDate = new Date();
+    switch (params.chartPeriod) {
+      case 'last2day-selector':
+        beginDate.setDate(beginDate.getDate() - 2);
+        break;
+      case 'last1week-selector':
+        beginDate.setDate(beginDate.getDate() - 7);
+        break;
+      case 'last1month-selector':
+        beginDate.setDate(beginDate.getDate() - 31);
+        break;
+      case 'last1year-selector':
+        beginDate.setDate(beginDate.getDate() - 365);
+        break;
+      case 'last1day-selector':
+      default:
+        beginDate.setDate(beginDate.getDate() - 1);
+    }
+
+    params.begin_date = beginDate;
+    params.end_date = new Date();
+
+    // Choose attributes
+    params.attributes_device = [];
+    params.attributes_device.push('name');
+    params.attributes_device.push('selector');
+    params.attributes_device_feature = [];
+    params.attributes_device_feature.push('id');
+    params.attributes_device_feature.push('name');
+    params.attributes_device_feature.push('selector');
+    params.attributes_device_feature.push('unit');
+    params.attributes_device_feature.push('last_value');
+    params.attributes_device_feature.push('last_value_changed');
+    params.attributes_device_feature.push('last_downsampling');
+    params.attributes_device_room = [];
+    params.attributes_device_room.push('name');
+    params.attributes_device_room.push('selector');
```

Comment:
This is weird. Why don't you just do: 

```
params.attributes_device = [
  'name',
  'selector'
];
```

Reply:
it's the list of attribute in select I need for display, I add this capability in device.get function and in device.getFeatureStates to reduce the size of data up from database

Reply:
Yes, but in your code:

```javascript
params.attributes_device_room = [];
    params.attributes_device_room.push('name');
    params.attributes_device_room.push('selector');
```

Is simply the exact same as:

```javascript
params.attributes_device_room = ['name', 'selector'];
```

Reply:
i fix this

Reply:
It's not fixed so I reopen it